### DISCUSSION TITLE
Run all unit test in all packages in repo

### DISCUSSION THREAD

**Discussion by markbeij** [codemagic-ci-cd/codemagic-docs#2014] [Q&A]
The `Flutter test` command only works for the current directory, but how can I run all unit tests in code magic CI/CD pipeline for my repo when I have multiple packages which have their own unit tests?
I'm not an expert in bash scripting but managed to loop through all test folders and run unit tests, but my build won't fail in on unit test failure. Is there a recommended way of doing this?

**Comment by himesh-cm** [comment]
Hi @markbeij, Please find here with the sample script.

```
for TEST_DIR in $(find . -name '*_test.dart' | xargs -n 1 dirname | sort -u); do
  # Run tests
  flutter test $TEST_DIR

  # Capture exit code
  if [ $? -ne 0 ]; then
    EXIT_CODE=1
  fi
done
```
The find command locates all files that match *_test.dart, which should find all your unit test files.

Feel free to modify the above script to meet your project requirements.

**Reply by markbeij** [reply]
Thank you for your help @himesh-cm! Unfortunately it's not working for me..
I get: 
```
Error: No pubspec.yaml file found.
This command should be run from the root of your Flutter project.
Error: No pubspec.yaml file found.
This command should be run from the root of your Flutter project.
Error: No pubspec.yaml file found.
This command should be run from the root of your Flutter project.
Error: No pubspec.yaml file found.
This command should be run from the root of your Flutter project.
```
Maybe it's because of our folder structure.. In our repo we have several apps and packages. All apps and packages live in their own folder, so it's like there is one pubspec.yaml for all of them. We have them in the same repository because they are all very small and are used together to form one product.

Thanks to your script I've created the following, but its not failing the build when a unit test fails
```bash
for TEST_DIR in $(find . -name "test" -not -path "*/.dart_tool/*" -not -path "*/testmocks/*" | sort -u); do
  # Go to parent folder of test folder, because that's where the pubspec.yaml is. Then run tests
  cd "$TEST_DIR/..";pwd;flutter test
  # Capture exit code
  if [ $? -ne 0 ]; then
    EXIT_CODE=1
    exit
  fi
  cd - #Go back to previous folder so the next iteration can run succesfully
done```

**Reply by himesh-cm** [reply]
@markbeij, Can you please share the build id, so i can look into it?

**Reply by markbeij** [reply]
64f72ce80c8e2fa7868c0bef

**Reply by himesh-cm** [reply]
You can use `exit 1` to end the workflow in your script.

**Reply by markbeij** [reply]
Thank you @himesh-cm, I'm a complete noob as it comes to bash scripting.. I'm not sure where to put the `exit 1`
Doing this within the loop makes the subprocess quit but not the main process.

It would've been great if CodeMagic has documentation on how they recommend doing this..

**Reply by himesh-cm** [reply]
In your if statement
```
if [ $? -ne 0 ]; then
    exit 1
  fi
```

**Comment by himesh-cm** [comment]
I'm tentatively closing the discussion since we haven't heard back from you. You can open the discussion again by replying to this message or opening a new discussion.
